File: Behavioural_Design_Pattern/Strategy_Pattern.java
package Behavioural_Design_Pattern;

abstract class PaymentStrategy {
    abstract void processPayment(double amount);
}

class CrediCard extends PaymentStrategy {
    void processPayment(double amount) {
        System.out.println("Payment of " + amount + "done by Credit Card");
    }
}

class UPI extends PaymentStrategy {
    void processPayment(double amount) {
        System.out.println("Payment of " + amount + "done by UPI");
    }
}

class PaymentProcessor {
    PaymentStrategy ps;

    PaymentProcessor(PaymentStrategy ps) {
        this.ps = ps;
    }

    void process(double amount) {
        ps.processPayment(amount);
    }
}

public class Strategy_Pattern {
    public static void main(String[] args) {
        PaymentStrategy credit = new CrediCard();
        PaymentProcessor pp = new PaymentProcessor(credit);
        pp.process(40);
    }
}

File: Behavioural_Design_Pattern/Strategy_Pattern.cpp
#include <bits/stdc++.h>
#include <iostream>
#include <string>
using namespace std;

class PaymentStrategy
{
public:
    virtual void processPayment(double amount) = 0;
    virtual ~PaymentStrategy() = default;
};
class CreditCard : public PaymentStrategy
{
public:
    void processPayment(double amount)
    {
        cout << "Payment of " << amount << "done by Credit Card" << endl;
    }
};
class UPI : public PaymentStrategy
{
public:
    void processPayment(double amount)
    {
        cout << "Payment of " << amount << "done by UPI" << endl;
    }
};
class PaymentProcessor
{
private:
    PaymentStrategy *ps;

public:
    PaymentProcessor(PaymentStrategy *ps)
    {
        this->ps = ps;
    }
    void process(double amount)
    {
        ps->processPayment(amount);
    }
};
int main()
{
    PaymentStrategy *creditCard = new CreditCard();
    PaymentStrategy *upi = new UPI();
    PaymentProcessor *ps1 = new PaymentProcessor(creditCard);
    PaymentProcessor *ps2 = new PaymentProcessor(upi);
    ps2->process(10000);
    ps1->process(300);
}